### ISSUE TITLE
Improve file saving at editor

### ISSUE THREAD

**Issue by aisbergde** [gsantner/markor#954]
#### General information

* **App version:** 
* **System:** 

#### Description

when editing a note and something happens with Markor (it could be stopped by Android, I don't know what happens) I want not lost my changed but not yet saved content.

The best would be to store content permanently in a hidden file copy and to allow me to restore when I reopen the file and Markor finds differences. Something like Microsoft Office programs do

An auto save option also could be fine in some cases, but this could also be unwanted in other cases. 

I just wrote on a note, then Marcor unexpected disappear. And after this the note file was empty.

**Comment by gsantner** [comment]
Every time I have encountered these "shadow saves" I was pretty disappointed. Half time it doesn't work well, you run into sync conflicts and stuff keeps asking over and over. Notably word and libreoffice. Thats not something I want to seek for to be honest.

Instead, I would opt for a single "security copy" in markor's cache folder, that just saves most recent opened file contents. Where you could go to and look it up.

Even that still, I'm not sure if it's a good idea. If it's something sensitive I would rather not have a copy somewhere around. Another issue, how to deal with encrypted files. As Markor also supports encryption, this also needs thoughts. Do save cleartext? Do save encrypted, and where to get password then?

A lot of open points, and frankly said I'm not very convinced of automated file duplicating.

**Comment by aisbergde** [comment]
I see, you are right, there are too many side effects. 

Maybe you could consider a button or a context menu option to quick enable and disable autosave by time. But this should be "per currently edit" option. Often I want to prevent some saving because I only read files, but when edit files, write notes, I would like to explicitly turn autosave on. The used time for this autosave could be set in settings.

**Comment by harshad1** [comment]
@aisbergde What are the disadvantages of auto-saving all the time? i.e. why does it need to be enabled and disabled? If you are viewing a file and making no edits then no 'save' will happen, right?

**Comment by gsantner** [comment]
(notice: auto saving happens only when you move to other app or close the editor, not during editing. And this is for the case you efficiently don't change anything but later on see a different file-modtime)

**Comment by aisbergde** [comment]
I have the issue only on my tablet. Often it closes apps (not only Markor) and I need to reopen. And in this situation I lost a longer collection of links and citation in the file I edited this time. Then I asked for this feature. But I will try to avoid taking notes on the tablet or I will learn to save more often. KISS is more important than auto save. It would be better to keep Markor as it is. I have not yet issues with unexpected app closing on the smartphone.

**Comment by gsantner** [comment]
OK then just for the record for myself: There is a **device specific hw/sw issue** that forcequits apps, leaving Markors open file unsaved.

**Comment by gsantner** [comment]
So I have thought multiple times on this:

What I don't want: Safety copies. And especially, no "unencrypted" copies of encrypted stuff.

What I prefer instead: Timer based saving, i.e. every minute. Up to decide deafulton-opt-out or defaultoff-opt-in etc.

**Comment by aisbergde** [comment]
time based saving (with an option to enable and disable and to set the time) could be fine.
BTW, this happend only once. So maybe it is even not very important. 

And I have no data loss issues on my smarthpone (only the issue, that "share with markor" often not creates an entry, when markor "sleeps", and need to check after each sharing)

I also changed the mode from "several instances of Markor" in "only one instance". Maybe this has an effect.

**Comment by korigod** [comment]
I've lost important notes several times with markor. I love the app so much and I don't want to use anythong else instead but this really bothers me, it's the single most important feature for the notes taking app, to not lose the notes! It looks like I have to keep pressing this Save button all the time or be afraid my notes will be lost. I was sure pressing Back button or switching screen away from Quick note saves too but it seems that's not the case (which is extremely strange).

As for shadow copies, I'm fine with them while I can choose their location. Any solution which will allow me not to press Save all the time and be certain I won't lose any data will be highly appreciated. Or at least move the save button to the panel from the menu if it's impossible to solve. :)

**Comment by seelaman** [comment]
> @aisbergde What are the disadvantages of auto-saving all the time? i.e. why does it need to be enabled and disabled? If you are viewing a file and making no edits then no 'save' will happen, right?

I synchronize text files between machines and use them through the day to track notes, tasks etc. (Markor and markdown in general is great for that).

I noticed that if my phone hasn't caught up syncing with my main machine and I open the file in Markor, the state Markor has is saved (locally on the phone). This state (which is missing updates I had done on my main machine earlier) is then synced back to the main machine, overwriting the work that happened there.

Autosave is definitely a good idea, but is it perhaps saving even when no changes have actually happened on the phone file Markor has open? If the last modified time of the file changes without there being an actual text diff, I could see why syncthing would assume that's the latest version.

**Comment by harshad1** [comment]
@seelaman Good point. I use foldersync + pcloud and have similar issues .. so I should have realized this.

**Comment by aisbergde** [comment]
@seelaman autosaving all the time is also OK if this only happens if text has been changed.

The option to disable autosave could be required, because some user want to control this.

For example when I work on the PC, for example with VSC, then I want to decide when my change explicitly is saved into the file. Until I save it explicitly, is is saved in something like a shadow copy and I like this behavior much more than a hard auto save.

**Comment by seelaman** [comment]
> @seelaman Good point. I use foldersync + pcloud and have similar issues .. so I should have realized this.

Interesting, I wasn't sure if this was Markor related or a syncthing issue but good to know you're facing similar issues with foldersync + pcloud

**Comment by PhilT** [comment]
I just hit this issue. Just finished typing and accidentally hit the top left corner of the screen where the folder name was. App crashed and lost the last 20 minutes of typing.

Perhaps a save prompt or autosave before any major actions would be useful

**Comment by gsantner** [comment]
There are more then enough discussions and issues about this around. closing duplicates.
